rfcs/0049-flakes.md

```diff
@@ -0,0 +1,703 @@
+mechanisms such as Nix channels and the Nix search path.
+
+# TLDR
+
+* A prototype implementation is available in Nixpkgs: `nix run
+  nixpkgs.nixFlakes`.
+
+* Flakes replace channels. For example,
+
+      # nix run nixpkgs:hello -c hello
+
+  fetches the latest version of [the `nixpkgs`
+  flake](https://github.com/edolstra/nixpkgs/blob/release-19.03/flake.nix)
+  from GitHub and builds its `hello` package. Similarly, Nix itself is
+  a flake, so you can get the latest version as follows:
+
+      # nix run nix
+
+* Flakes are looked up in a
+  [registry](https://raw.githubusercontent.com/NixOS/flake-registry/master/flake-registry.json)
+  that maps identifiers such as `nixpkgs` to actual locations such as
+  `github:edolstra/nixpkgs/release-19.03`. You can use such locations
+  ("flake references") directly in the `nix` command:
+
+  ```
+  # nix build github:NixOS/patchelf
+  ```
+
+* For a reproducible result, you can also use a specific revision:
+
+  ```
+  # nix build nixpkgs/a0e1f50e6f72e5037d71a0b65c67cf0605349a06:hello
+  ```
+
+* To get information about a flake:
+
+  ```
+  # nix flake info nixpkgs
+  Description:   A collection of packages for the Nix package manager
+  Revision:      a0e1f50e6f72e5037d71a0b65c67cf0605349a06
+  ...
+  ```
+
+* In addition to the global registry, there is a per-user
+  registry. This can be used to pin flakes to the current version:
+  ```
+  # nix flake pin nixpkgs
+  ```
+  or to a specific version:
+  ```
+  # nix flake add nixpkgs github:edolstra/nixpkgs/a0e1f50e6f72e5037d71a0b65c67cf0605349a06
+  ```
+
+* Flakes can have dependencies on other flakes. For example, [the
+  `patchelf`
+  flake](https://github.com/NixOS/patchelf/blob/master/flake.nix)
+  depends on the `nixpkgs` flake. To ensure reproducibility,
+  dependencies are pinned to specific versions using a *lock file*
+  (e.g. [for
+  patchelf](https://github.com/NixOS/patchelf/blob/master/flake.lock)). Lock
+  files are generated automatically.
+
+* Flakes are evaluated in pure mode, meaning they can't access
+  anything other than their own source or declared dependencies. This
+  allows `nix` to cache evaluation results.
+
+* The `nix` command is flake-based. For example,
+  ```
+  # nix build
+  ```
+  builds the flake in the current directory.
+
+# Motivation
+[motivation]: #motivation
+
+Flakes are motivated by a number of serious shortcomings in Nix:
+
+* While Nix pioneered reproducible builds, sadly, Nix expressions are
+  not nearly as reproducible as Nix builds. Nix expressions can access
+  arbitrary files (such as `~/.config/nixpkgs/config.nix`),
+  environment variables, and Git repositories. This means for instance
+  that it is hard to ensure reproducible evaluation of NixOS or NixOps
+  configurations.
+
+* Nix projects lack discoverability and a standard structure. For
+  example, it's just convention that a repository has a `release.nix`
+  for Hydra jobs and a `default.nix` for packages.
```

Comment:
Isn't `default.nix` and `shell.nix` assumed by enough tools to be a bit more than mere convention?

Reply:
No, `default.nix` has no well-known structure. I've seen it contain a single package at top-level, an attrset of packages, and an attrset of arbitrary stuff (like functions).